File: src/lib/auth/auth-service.ts
import bcrypt from 'bcryptjs';
import jwt from 'jsonwebtoken';
import crypto from 'crypto';
import { UserRepository } from '../repositories/user-repository';
import { User, Session } from '../models/types';

// JWT secret key from environment variables
const JWT_SECRET = process.env.JWT_SECRET || 'default-secret-change-in-production';
const JWT_EXPIRES_IN = process.env.JWT_EXPIRES_IN || '7d';
const BCRYPT_SALT_ROUNDS = 12;

/**
 * Authentication service for user management and authentication
 */
export class AuthService {
  private userRepository: UserRepository;

  constructor() {
    this.userRepository = new UserRepository();
  }

  /**
   * Register a new user
   * @param email User email
   * @param password User password
   * @param name User name (optional)
   * @returns Created user object (without password)
   * @throws Error if email already exists
   */
  async register(email: string, password: string, name?: string): Promise<Omit<User, 'password_hash'>> {
    // Check if user already exists
    const existingUser = await this.userRepository.findByEmail(email);
    if (existingUser) {
      throw new Error('Email already registered');
    }

    // Hash password
    const passwordHash = await bcrypt.hash(password, BCRYPT_SALT_ROUNDS);

    // Create user
    const user = await this.userRepository.create({
      email,
      password_hash: passwordHash,
      name: name || null,
      role: 'user',
      email_verified: false,
    });

    // Return user without password hash
    const { password_hash, ...userWithoutPassword } = user;
    return userWithoutPassword;
  }

  /**
   * Login a user
   * @param email User email
   * @param password User password
   * @returns JWT token and user object (without password)
   * @throws Error if credentials are invalid
   */
  async login(email: string, password: string): Promise<{ token: string; user: Omit<User, 'password_hash'> }> {
    // Find user by email
    const user = await this.userRepository.findByEmail(email);
    if (!user) {
      throw new Error('Invalid credentials');
    }

    // Verify password
    const isPasswordValid = await bcrypt.compare(password, user.password_hash);
    if (!isPasswordValid) {
      throw new Error('Invalid credentials');
    }

    // Generate JWT token
    const token = this.generateToken(user);

    // Create session
    const sessionExpires = new Date();
    sessionExpires.setDate(sessionExpires.getDate() + 7); // 7 days from now

    await this.userRepository.createSession({
      user_id: user.id,
      token_hash: this.hashToken(token),
      expires_at: sessionExpires,
    });

    // Return token and user without password hash
    const { password_hash, ...userWithoutPassword } = user;
    return { token, user: userWithoutPassword };
  }

  /**
   * Verify a JWT token
   * @param token JWT token
   * @returns User ID if token is valid
   * @throws Error if token is invalid
   */
  async verifyToken(token: string): Promise<number> {
    try {
      // Verify JWT signature
      const decoded = jwt.verify(token, JWT_SECRET) as { userId: number };
      
      // Check if session exists
      const session = await this.userRepository.findSessionByToken(this.hashToken(token));
      if (!session) {
        throw new Error('Session not found');
      }

      return decoded.userId;
    } catch (error) {
      throw new Error('Invalid token');
    }
  }

  /**
   * Get user by ID
   * @param userId User ID
   * @returns User object (without password) or null if not found
   */
  async getUserById(userId: number): Promise<Omit<User, 'password_hash'> | null> {
    const user = await this.userRepository.findById(userId);
    if (!user) {
      return null;
    }

    // Return user without password hash
    const { password_hash, ...userWithoutPassword } = user;
    return userWithoutPassword;
  }

  /**
   * Logout a user by invalidating their session
   * @param token JWT token
   * @returns True if logged out successfully
   */
  async logout(token: string): Promise<boolean> {
    try {
      const session = await this.userRepository.findSessionByToken(this.hashToken(token));
      if (session) {
        await this.userRepository.deleteSession(session.id);
        return true;
      }
      return false;
    } catch (error) {
      return false;
    }
  }

  /**
   * Change user password
   * @param userId User ID
   * @param currentPassword Current password
   * @param newPassword New password
   * @returns True if password changed successfully
   * @throws Error if current password is invalid
   */
  async changePassword(userId: number, currentPassword: string, newPassword: string): Promise<boolean> {
    // Find user
    const user = await this.userRepository.findById(userId);
    if (!user) {
      throw new Error('User not found');
    }

    // Verify current password
    const isPasswordValid = await bcrypt.compare(currentPassword, user.password_hash);
    if (!isPasswordValid) {
      throw new Error('Current password is incorrect');
    }

    // Hash new password
    const newPasswordHash = await bcrypt.hash(newPassword, BCRYPT_SALT_ROUNDS);

    // Update user
    await this.userRepository.update(userId, { password_hash: newPasswordHash });

    // Invalidate all sessions
    await this.userRepository.deleteUserSessions(userId);

    return true;
  }

  /**
   * Create a password reset token
   * @param email User email
   * @returns Reset token or null if user not found
   */
  async createPasswordResetToken(email: string): Promise<string | null> {
    // Find user by email
    const user = await this.userRepository.findByEmail(email);
    if (!user) {
      return null;
    }

    // Generate token
    const resetToken = crypto.randomBytes(32).toString('hex');
    const tokenHash = this.hashToken(resetToken);

    // Set expiration (1 hour)
    const expires = new Date();
    expires.setHours(expires.getHours() + 1);

    // Save token
    await this.userRepository.createPasswordResetToken({
      user_id: user.id,
      token_hash: tokenHash,
      expires_at: expires,
    });

    return resetToken;
  }

  /**
   * Reset password using a reset token
   * @param token Reset token
   * @param newPassword New password
   * @returns True if password reset successfully
   * @throws Error if token is invalid
   */
  async resetPassword(token: string, newPassword: string): Promise<boolean> {
    // Find token
    const resetToken = await this.userRepository.findPasswordResetToken(this.hashToken(token));
    if (!resetToken) {
      throw new Error('Invalid or expired token');
    }

    // Hash new password
    const newPasswordHash = await bcrypt.hash(newPassword, BCRYPT_SALT_ROUNDS);

    // Update user
    await this.userRepository.update(resetToken.user_id, { password_hash: newPasswordHash });

    // Delete token
    await this.userRepository.deletePasswordResetToken(resetToken.id);

    // Invalidate all sessions
    await this.userRepository.deleteUserSessions(resetToken.user_id);

    return true;
  }

  /**
   * Clean up expired sessions and tokens
   */
  async cleanup(): Promise<void> {
    await this.userRepository.cleanupExpiredSessions();
    await this.userRepository.cleanupExpiredPasswordResetTokens();
  }

  /**
   * Generate a JWT token for a user
   * @param user User object
   * @returns JWT token
   */
  private generateToken(user: User): string {
    return jwt.sign({ userId: user.id, role: user.role }, JWT_SECRET, {
      expiresIn: JWT_EXPIRES_IN,
    });
  }

  /**
   * Hash a token for storage
   * @param token Token to hash
   * @returns Hashed token
   */
  private hashToken(token: string): string {
    return crypto.createHash('sha256').update(token).digest('hex');
  }
}

File: venue-explorer-api/app/services/auth_service.py
"""
Authentication service for user management and JWT token handling
"""

import bcrypt
import jwt
from datetime import datetime, timedelta, timezone
from typing import Optional, Dict, Any
from sqlalchemy.orm import Session
from sqlalchemy.exc import IntegrityError

from app.models.orm import User, UserFavorite
from app.schemas.models import UserCreate, UserLogin, UserResponse, TokenResponse, FavoriteRequest
from app.core.config import Settings


class AuthService:
    """Authentication service for user registration, login, and token management"""
    
    def __init__(self, db: Session, settings: Settings):
        self.db = db
        self.settings = settings
        self.secret_key = settings.SECRET_KEY
        self.algorithm = "HS256"
        self.access_token_expire_minutes = 60 * 24 * 7  # 7 days
        self.bcrypt_salt_rounds = 12
    
    async def register(self, user_data: UserCreate) -> Dict[str, Any]:
        """
        Register a new user
        
        Args:
            user_data: User registration data
            
        Returns:
            Dictionary containing user data and access token
            
        Raises:
            ValueError: If email already exists or validation fails
        """
        # Check if user already exists
        existing_user = self.db.query(User).filter(User.email == user_data.email).first()
        if existing_user:
            raise ValueError("Email already registered")
        
        # Hash password
        password_hash = bcrypt.hashpw(
            user_data.password.encode('utf-8'), 
            bcrypt.gensalt(rounds=self.bcrypt_salt_rounds)
        ).decode('utf-8')
        
        # Create user
        try:
            user = User(
                email=user_data.email,
                password_hash=password_hash,
                name=user_data.name,
                role='user',
                email_verified=False
            )
            self.db.add(user)
            self.db.commit()
            self.db.refresh(user)
        except IntegrityError:
            self.db.rollback()
            raise ValueError("Email already registered")
        
        # Generate access token
        access_token = self._create_access_token({"sub": str(user.id), "email": user.email})
        
        return {
            "user": UserResponse.from_orm(user),
            "token": TokenResponse(
                access_token=access_token,
                expires_in=self.access_token_expire_minutes * 60
            )
        }
    
    async def login(self, login_data: UserLogin) -> Dict[str, Any]:
        """
        Authenticate user and return access token
        
        Args:
            login_data: User login credentials
            
        Returns:
            Dictionary containing user data and access token
            
        Raises:
            ValueError: If credentials are invalid
        """
        # Find user by email
        user = self.db.query(User).filter(User.email == login_data.email).first()
        if not user:
            raise ValueError("Invalid email or password")
        
        # Verify password
        if not bcrypt.checkpw(login_data.password.encode('utf-8'), user.password_hash.encode('utf-8')):
            raise ValueError("Invalid email or password")
        
        # Generate access token
        access_token = self._create_access_token({"sub": str(user.id), "email": user.email})
        
        return {
            "user": UserResponse.from_orm(user),
            "token": TokenResponse(
                access_token=access_token,
                expires_in=self.access_token_expire_minutes * 60
            )
        }
    
    async def get_current_user(self, token: str) -> Optional[User]:
        """
        Get current user from JWT token
        
        Args:
            token: JWT access token
            
        Returns:
            User object if token is valid, None otherwise
        """
        try:
            payload = jwt.decode(token, self.secret_key, algorithms=[self.algorithm])
            user_id = payload.get("sub")
            if user_id is None:
                return None
            
            user = self.db.query(User).filter(User.id == int(user_id)).first()
            return user
        except (jwt.PyJWTError, ValueError):
            return None
    
    async def refresh_token(self, token: str) -> Optional[TokenResponse]:
        """
        Refresh an access token
        
        Args:
            token: Current JWT access token
            
        Returns:
            New token response if valid, None otherwise
        """
        user = await self.get_current_user(token)
        if not user:
            return None
        
        access_token = self._create_access_token({"sub": str(user.id), "email": user.email})
        return TokenResponse(
            access_token=access_token,
            expires_in=self.access_token_expire_minutes * 60
        )
    
    async def add_favorite(self, user_id: int, favorite_data: FavoriteRequest) -> Dict[str, Any]:
        """
        Add item to user favorites
        
        Args:
            user_id: User ID
            favorite_data: Favorite item data
            
        Returns:
            Success message
            
        Raises:
            ValueError: If favorite already exists
        """
        # Check if favorite already exists
        existing_favorite = self.db.query(UserFavorite).filter(
            UserFavorite.user_id == user_id,
            UserFavorite.entity_type == favorite_data.entity_type,
            UserFavorite.entity_id == favorite_data.entity_id
        ).first()
        
        if existing_favorite:
            raise ValueError("Item already in favorites")
        
        # Create favorite
        favorite = UserFavorite(
            user_id=user_id,
            entity_type=favorite_data.entity_type,
            entity_id=favorite_data.entity_id
        )
        self.db.add(favorite)
        self.db.commit()
        
        return {"message": "Item added to favorites successfully"}
    
    async def remove_favorite(self, user_id: int, entity_type: str, entity_id: int) -> Dict[str, Any]:
        """
        Remove item from user favorites
        
        Args:
            user_id: User ID
            entity_type: Type of entity (venue, artist)
            entity_id: Entity ID
            
        Returns:
            Success message
            
        Raises:
            ValueError: If favorite doesn't exist
        """
        favorite = self.db.query(UserFavorite).filter(
            UserFavorite.user_id == user_id,
            UserFavorite.entity_type == entity_type,
            UserFavorite.entity_id == entity_id
        ).first()
        
        if not favorite:
            raise ValueError("Item not found in favorites")
        
        self.db.delete(favorite)
        self.db.commit()
        
        return {"message": "Item removed from favorites successfully"}
    
    async def get_user_favorites(self, user_id: int) -> Dict[str, Any]:
        """
        Get user's favorite items
        
        Args:
            user_id: User ID
            
        Returns:
            Dictionary containing favorite venues and artists
        """
        favorites = self.db.query(UserFavorite).filter(UserFavorite.user_id == user_id).all()
        
        venues = []
        artists = []
        
        for favorite in favorites:
            if favorite.entity_type == 'venue':
                venues.append(favorite.entity_id)
            elif favorite.entity_type == 'artist':
                artists.append(favorite.entity_id)
        
        return {
            "venues": venues,
            "artists": artists,
            "total": len(favorites)
        }
    
    async def update_user_profile(self, user_id: int, name: Optional[str] = None) -> UserResponse:
        """
        Update user profile information
        
        Args:
            user_id: User ID
            name: New name (optional)
            
        Returns:
            Updated user response
            
        Raises:
            ValueError: If user not found
        """
        user = self.db.query(User).filter(User.id == user_id).first()
        if not user:
            raise ValueError("User not found")
        
        if name is not None:
            user.name = name
        
        user.updated_at = datetime.utcnow()
        self.db.commit()
        self.db.refresh(user)
        
        return UserResponse.from_orm(user)
    
    async def change_password(self, user_id: int, current_password: str, new_password: str) -> Dict[str, Any]:
        """
        Change user password
        
        Args:
            user_id: User ID
            current_password: Current password
            new_password: New password
            
        Returns:
            Success message
            
        Raises:
            ValueError: If current password is invalid or user not found
        """
        user = self.db.query(User).filter(User.id == user_id).first()
        if not user:
            raise ValueError("User not found")
        
        # Verify current password
        if not bcrypt.checkpw(current_password.encode('utf-8'), user.password_hash.encode('utf-8')):
            raise ValueError("Current password is incorrect")
        
        # Hash new password
        new_password_hash = bcrypt.hashpw(
            new_password.encode('utf-8'), 
            bcrypt.gensalt(rounds=self.bcrypt_salt_rounds)
        ).decode('utf-8')
        
        user.password_hash = new_password_hash
        user.updated_at = datetime.utcnow()
        self.db.commit()
        
        return {"message": "Password changed successfully"}
    
    async def verify_email(self, user_id: int) -> Dict[str, Any]:
        """
        Mark user email as verified
        
        Args:
            user_id: User ID
            
        Returns:
            Success message
            
        Raises:
            ValueError: If user not found
        """
        user = self.db.query(User).filter(User.id == user_id).first()
        if not user:
            raise ValueError("User not found")
        
        user.email_verified = True
        user.updated_at = datetime.utcnow()
        self.db.commit()
        
        return {"message": "Email verified successfully"}
    
    def _create_access_token(self, data: Dict[str, Any]) -> str:
        """
        Create JWT access token
        
        Args:
            data: Token payload data
            
        Returns:
            JWT token string
        """
        to_encode = data.copy()
        expire = datetime.now(timezone.utc) + timedelta(minutes=self.access_token_expire_minutes)
        to_encode.update({
            "exp": expire,
            "iat": datetime.now(timezone.utc),
            "type": "access"
        })
        encoded_jwt = jwt.encode(to_encode, self.secret_key, algorithm=self.algorithm)
        return encoded_jwt
    
    def _verify_token(self, token: str) -> Optional[Dict[str, Any]]:
        """
        Verify and decode JWT token
        
        Args:
            token: JWT token string
            
        Returns:
            Decoded payload if valid, None otherwise
        """
        try:
            payload = jwt.decode(token, self.secret_key, algorithms=[self.algorithm])
            return payload
        except jwt.PyJWTError:
            return None
